### PR TITLE
Read config metabucket

### PR DESCRIPTION
**Pull request by dvaldivia**
the `Meta` module will asynchronously read the configurations from the meta bucket and load it into the shared state `Config`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,6 +21,19 @@ To run the project you may specify a `.toml` configuration file, if none is spec
...
 A sample configuration file can be found at `config.toml.template`
 
+### Meta Bucket
+
+A bucket to store shared configuration across multiple `MinSQL` instances is needed, the location and access to it should be configured on the `config.toml`.
+
+````toml
+[server]
+   ...
+   metadata_bucket = "meta"
```
Comment (harshavardhana) [inline README.md]: This names too generic, prefix this with MinSQL? Like `minsql-meta`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (harshavardhana) [reply README.md]: Also create a bucket automatically if it doesn't exist? Such as default to `minsql-meta` if not specified ?

Comment (dvaldivia) [reply README.md]: you mean the name of the bucket? It's a configuration so they can choose the name of the bucket on their own, the example is meta.
Also we can definitively make it so it auto creates the meta bucket, if it exists, for that purpose choosing a default meta bucket name makes sense, however if they are running multiple MinSQL clusters, maybe we want them to think a little bit about the meta bucket name.

Comment (harshavardhana) [reply README.md]: Even when we document we should say `minsql-meta` most users will choose `meta` which is too generic, or specify how they should choose a bucket based on multiple MinSQL clusters.

Comment (dvaldivia) [reply README.md]: that makes sense

File: src/meta.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+pub struct Meta {
+    config: Arc<RwLock<Config>>,
+}
+
+impl Meta {
+    pub fn new(cfg: Arc<RwLock<Config>>) -> Meta {
+        Meta { config: cfg }
+    }
+
+    /// Scans the metabucket for configuration files and loads them into the shard state `Config`
```
Comment (krisis) [inline src/meta.rs]:
```suggestion
    /// Scans the metabucket for configuration files and loads them into the shared state `Config`
```

File: README.md (hunk 1 above)
Comment (dvaldivia) [reply README.md]: I did the change btw

File: src/storage.rs [hunk 3]
```diff
@@ -122,8 +125,13 @@ pub fn client_for_datastore(datastore: &DataStore) -> S3Client {
     S3Client::new_with(dispatcher, provider, region)
 }
 
-// <p>Function used to verify if a datastore is valid in terms of reachability</p>
-pub fn can_reach_datastore(datastore: &DataStore) -> bool {
+pub enum ReachableDatastoreError {
+    NoSuchBucket(String),
+    Unknown,
```
Comment (krisis) [inline src/storage.rs]: Could we have a String added to the Unknown variant to indicate what the error the S3 backend returned?
Like, 
```suggestion
    Unknown(String),
```

File: src/meta.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                                        error!("concatenating body: {:?}", e);
+                                        ()
+                                    })
+                                    .map(move |bytes| {
+                                        let result = String::from_utf8(bytes).unwrap();
+                                        let key: String = file_key_clone
+                                            .chars()
+                                            .skip("minsql/meta/".len())
+                                            .collect();
+                                        let parts: Vec<&str> = key.split("/").collect();
```
Comment (krisis) [inline src/meta.rs]: suggestion: you could get `parts` simply as below
```rust
let parts: Vec<&str> = file_key_clone.trim_start_matches("myminio/meta/").split("/").collect()
```

**Review by krisis — CHANGES_REQUESTED** [review]

File: src/storage.rs (hunk 3 above)
Comment (dvaldivia) [reply src/storage.rs]: ready, I improved the way we wrap around the errors so it's more future ready

File: README.md [hunk 5]
```diff
@@ -21,6 +21,19 @@ To run the project you may specify a `.toml` configuration file, if none is spec
 ```
 A sample configuration file can be found at `config.toml.template`
 
+### Meta Bucket
```
Comment (donatello) [inline README.md:24]: Configuration or System bucket seems to be more appropriate as we are not actually storing any metadata about data ingested by minsql

File: src/lib.rs [hunk 6]
```diff
@@ -231,9 +240,24 @@ impl MinSQL {
...
-                process::exit(0x0100);
+            match storage::can_reach_datastore(&ds) {
+                Ok(true) => (),
+                Ok(false) => {
+                    println!("{} datastore is not reachable", ds_name);
+                    process::exit(0x0100);
+                }
+                Err(e) => match e {
+                    storage::StorageError::Operation(ope) => match ope {
+                        storage::ReachableDatastoreError::NoSuchBucket(s) => {
```
Comment (donatello) [inline src/lib.rs]: Is this the only possibility for this match block?

File: src/meta.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                prefix: Some("minsql/meta/".to_owned()),
+                ..Default::default()
+            })
+            .map(|list_objects| list_objects.contents.unwrap())
+            .map_err(|_| ())
+            .and_then(|objects| {
+                // For each objects, get_object, filter out system files
+                stream::iter_ok(objects)
+                    .map(|file_object| file_object.clone().key.unwrap())
+                    .filter(|file_key| file_key.ends_with(".DS_Store") == false)
```
Comment (donatello) [inline src/meta.rs]:
```suggestion               
```

File: src/meta.rs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                                // Deserialize the object output and wrap in an `MetaConfigObject`
+                                object_output
+                                    .body
+                                    .unwrap()
+                                    .concat2()
+                                    .map_err(|e| {
+                                        error!("concatenating body: {:?}", e);
+                                        ()
+                                    })
+                                    .map(move |bytes| {
```
Comment (donatello) [inline src/meta.rs:121]: The content of this map function should be a separate function.

File: src/meta.rs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                        MetaConfigObject::Log(l) => {
+                            cfg.log.insert(l.clone().name.unwrap(), l);
+                        }
+                        MetaConfigObject::DataStore(ds) => {
+                            cfg.datastore.insert(ds.clone().name.unwrap(), ds);
+                        }
+                        _ => (),
+                    }
+                }
+            })
```
Comment (donatello) [inline src/meta.rs:166]: This map also is better separated out into a function.

**Review by donatello — CHANGES_REQUESTED** [review]

File: src/meta.rs (hunk 9 above)
Comment (dvaldivia) [reply src/meta.rs:166]: It will be refactored but there's still improvements that need to be made on top of this pipeline

Comment (donatello) [reply src/meta.rs:166]: Ok.

File: src/lib.rs (hunk 6 above)
Comment (donatello) [reply src/lib.rs]: What about this one? Only one case in the match block?

File: README.md (hunk 5 above)
Comment (donatello) [reply README.md:24]: We'll wait to discuss with @abperiasamy

**Pull request closed without merging**